### ISSUE TITLE
--terragrunt-source does not work consistently over plan and plan-all

### ISSUE THREAD

**Issue by TheKangaroo** [gruntwork-io/terragrunt#1057]
We use a module source like
```
source = "git::git@gitlab.example.com:test/tf-modules/mymodule.git?ref=master"
```
The tf files in the module are on the top level and the module does not include any subdirs.
I can run 
```
terragrunt plan --terragrunt-source ../../../../../tf-modules/mymodule
```
and it will successfully build the cache and run the plan.
In this case, all the files from my terragrunt repo and all the files from my module are placed directly in the .terragrunt-cach/xxxxx/xxxx/ folder.
When I run the same command but with plan-all my run will fail with tf files not found.
```
terragrunt plan-all --terragrunt-source ../../../../../tf-modules/mymodule
```
This will copy my module files to .terragrunt-cach/xxxxx/xxxx/ but places the terragrunt files under .terragrunt-cach/xxxxx/xxxx/mymodule/ . The mymodule subfolder seems to be the working directory in this case, and therefore the tf files are not present there.
Another option that works with plan-all is
```
terragrunt plan-all --terragrunt-source ../../../../../tf-modules 
(omitting the module folder <mymodule> itself)
```
This copies everithing into the .terragrunt-cach/xxxxx/xxxx/mymodule/ sub-folder which works, but is diffrent from the behavior of just planing one terragrunt folder.

I think this problem occurs due to my tf files at the root level in the git repo and therefore no double-slash (//) in my source path.
I can omit this problem if I add a `//.` to my git source.
```
source = "git::git@gitlab.example.com:test/tf-modules/mymodule.git//.?ref=master"
```
But I think this should work without specifying the root path at the module level.

**Comment by yorinasub17** [comment]
The double slash is necessary for `xxx-all` commands because it is a multi-module function. As such, terragrunt needs to know which part of the source to replace so that it can replace the source for multiple modules.

This behavior is documented in https://terragrunt.gruntwork.io/docs/features/execute-terraform-commands-on-multiple-modules-at-once/#testing-multiple-modules-locally

**Comment by TheKangaroo** [comment]
Ah, I see, thank you @yorinasub17 for the explanation and the link.
First I'm still not 100% sure how to handle this in my case. All the examples I saw (like in the link you provided) assume I have a subpath in my git repo to the actual module. But my git repo ist just a flat structure of a few tf files.  Am I missing some best practices to put my tf files in a subfolder? Or is it just fine to have a repo without subfolders and I should add the `//.` to my path like
```
source = "git::git@gitlab.example.com:test/tf-modules/mymodule.git//.?ref=master"
```
?

And second, I don't really understand why apply (without -all) behaves different from this. 

Sorry if this are silly questions, but I'm new to terraform and terragrunt and try to understand the mechanisms behind it.

**Comment by yorinasub17** [comment]
> Or is it just fine to have a repo without subfolders and I should add the //. to my path like

Yes this is basically how to handle this and still support the multi module source behavior.

> And second, I don't really understand why apply (without -all) behaves different from this.

This is indeed confusing, but the way to understand this is that in `apply` (or any command that is not `xxx-all`), you are dealing with a single module. Since terragrunt knows this is a single module command, it knows it is safe to verbatim update the `terraform.source` property with `--terragrunt-source`.

However, with `xxx-all`, it is a multi module command and terragrunt knows that. Since it is dealing with multiple modules, it doesn't make sense for terragrunt to go in and verbatim update `terraform.source` on all the modules it finds, since that means it runs the same module with different inputs. So instead, it uses the `//` as a reference point for knowing what part of the path to update with the `--terragrunt-source` parameter.

With that said, there may be a bug here where it doesn't correctly handle the case when there is no `//` and root repos. This isn't a use case we use heavily, so we most likely won't get into resolving this issue, but if anyone from the community is available a PR would be very welcome and appreciated here.
